index.js

```diff
@@ -47,8 +46,12 @@ function paginate(query, options, callback) {
         docs:  Promise.resolve([]),
         count: (() => {
             return new Promise((resolve) => {
-                this.find(query).select({ _id: 1 }).sort(sort).limit(totalLimit).lean(true).exec().then((items => {
-                    return resolve(items.length)
+                this.find(query).select({ _id: 1 }).sort(sort).skip(totalLimit - 1).limit(1).lean(true).exec().then((items => {
```

Comment:
Oh. understand...
"limit(totalLimit)" : move position contains all of data
"skip(totalLimit - 1).limit(1)" : move position not contains all of data.

Reply:
If I understand right,
`this.find(query).select({ _id: 1 }).sort(sort).skip(totalLimit - 1).limit(1).lean(true).exec()` will jump to `totalLimit -1` position and take **1 item**. But
`this.find(query).select({ _id: 1 }).sort(sort).limit(totalLimit).lean(true).exec()` will take **totalLimit items** from the beginning.
So, their results should be different. If we use this to check there are any more items after `totalLimit`, it's ok. But if we use it for getting **total number of items**, I think the new code works more correctly.
I just curious should we need to take the items in this case if we only need the count? Can we just count like this? **count()** queries from metadata so it should be faster than counting in index.
`Math.min(totalLimit, await this.find(query).count())`

Reply:
- For your understanding,
- Reid and I decided to give all request the limit after discussing pagination performance. (Actually, we don't need total count for pagination.) We already checked how to use feature including pagination from OPS except for order search in LH. So, just 1 API is using **_totalLimit_** option and Reid seems to avoid scan data for the count.

Reply:
I understand what you mean, but as I know `count()` will use indexes if the condition provided. This code was a workaround of limiting the `count()` or `countDocuments()` to avoid traversing the large collection but I found the better solution that they have a `limit` option. I didn't know that they provide this option.

As the `count()` is deprecated, I'm going to use `countDocuments()` instead like below:
```
this.countDocuments(query, { limit: totalLimit })
```
```

Reply:
It looks like mongoose doesn't support the limit for `count` or `countDocuments`.

Reply:
Please give me your opinions for this workaround.

```
                this.find(query).select({ _id: 1 }).sort(sort).skip(totalLimit - 1).limit(1).lean(true).exec().then((items => {
                    if (items.length > 0) {
                        return resolve(totalLimit)
                    } else {
                        return resolve(this.countDocuments(query).exec().then(count => count))
                    }
                }))
```

The result is like below:
```
// Greater than totalLimit
mongo_1  | 2020-04-16T04:31:48.168+0000 I  COMMAND  [conn223] command baemin.orders command: find { find: "orders", filter: { $or: [ { status: "READY" }, { status: "ASSIGNED" }, { status: "PICKED" } ] }, sort: { createdAt: -1 }, projection: { id: 1, csId: 1, userId: 1, merchantId: 1, agentId: 1, dishes: 1, isPartner: 1, canAcceptOrder: 1, allowCashlessTransaction: 1, payment: 1, fromName: 1, fromPhone: 1, fromAddress: 1, fromImageUrl: 1, toName: 1, toPhone: 1, toAddress: 1, toAddressDetail: 1, status: 1, discountedTotalPrice: 1, createdAt: 1, updatedAt: 1 }, limit: 30, returnKey: false, showRecordId: false, lsid: { id: UUID("907f7e45-572d-494b-8b4d-28290018ad62") }, $db: "baemin" } planSummary: IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: 1 } keysExamined:32 docsExamined:30 cursorExhausted:1 numYields:7 nreturned:30 queryHash:E4B4012E planCacheKey:CACB69CA reslen:36898 locks:{ ReplicationStateTransition: { acquireCount: { w: 8 } }, Global: { acquireCount: { r: 8 } }, Database: { acquireCount: { r: 8 } }, Collection: { acquireCount: { r: 8 } }, Mutex: { acquireCount: { r: 1 } } } storage:{} protocol:op_query 10ms
mongo_1  | 2020-04-16T04:31:48.168+0000 I  COMMAND  [conn232] command baemin.orders command: find { find: "orders", filter: { $or: [ { status: "READY" }, { status: "ASSIGNED" }, { status: "PICKED" } ] }, sort: { createdAt: -1 }, projection: { _id: 1 }, skip: 3999, limit: 1, returnKey: false, showRecordId: false, lsid: { id: UUID("ae9a6090-6c5d-4239-a98f-95d4dd6946c3") }, $db: "baemin" } planSummary: IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: 1 } keysExamined:697 docsExamined:0 cursorExhausted:1 numYields:18 nreturned:0 queryHash:1FF13A0E planCacheKey:79CA1331 reslen:117 locks:{ ReplicationStateTransition: { acquireCount: { w: 19 } }, Global: { acquireCount: { r: 19 } }, Database: { acquireCount: { r: 19 } }, Collection: { acquireCount: { r: 19 } }, Mutex: { acquireCount: { r: 1 } } } storage:{} protocol:op_query 11ms
mongo_1  | 2020-04-16T04:31:48.211+0000 I  COMMAND  [conn224] command baemin.orders command: aggregate { aggregate: "orders", pipeline: [ { $match: { $or: [ { status: "READY" }, { status: "ASSIGNED" }, { status: "PICKED" } ] } }, { $group: { _id: null, n: { $sum: 1 } } } ], cursor: {}, lsid: { id: UUID("ae9a6090-6c5d-4239-a98f-95d4dd6946c3") }, $db: "baemin" } planSummary: IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: -1 } keysExamined:698 docsExamined:0 cursorExhausted:1 numYields:12 nreturned:1 queryHash:F75B98C2 planCacheKey:2C50530F reslen:137 locks:{ ReplicationStateTransition: { acquireCount: { w: 14 } }, Global: { acquireCount: { r: 14 } }, Database: { acquireCount: { r: 14 } }, Collection: { acquireCount: { r: 14 } }, Mutex: { acquireCount: { r: 2 } } } storage:{} protocol:op_query 5ms

// Less than totalLimit
mongo_1  | 2020-04-16T04:37:53.090+0000 I  COMMAND  [conn229] command baemin.orders command: find { find: "orders", filter: { $or: [ { status: "READY" }, { status: "ASSIGNED" }, { status: "PICKED" }, { status: "CANCELED" } ] }, sort: { createdAt: -1 }, projection: { id: 1, csId: 1, userId: 1, merchantId: 1, agentId: 1, dishes: 1, isPartner: 1, canAcceptOrder: 1, allowCashlessTransaction: 1, payment: 1, fromName: 1, fromPhone: 1, fromAddress: 1, fromImageUrl: 1, toName: 1, toPhone: 1, toAddress: 1, toAddressDetail: 1, status: 1, discountedTotalPrice: 1, createdAt: 1, updatedAt: 1 }, limit: 30, returnKey: false, showRecordId: false, lsid: { id: UUID("907f7e45-572d-494b-8b4d-28290018ad62") }, $db: "baemin" } planSummary: IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: 1 } keysExamined:33 docsExamined:30 cursorExhausted:1 numYields:9 nreturned:30 queryHash:93D550FC planCacheKey:79896B46 reslen:31141 locks:{ ReplicationStateTransition: { acquireCount: { w: 10 } }, Global: { acquireCount: { r: 10 } }, Database: { acquireCount: { r: 10 } }, Collection: { acquireCount: { r: 10 } }, Mutex: { acquireCount: { r: 1 } } } storage:{} protocol:op_query 12ms
mongo_1  | 2020-04-16T04:37:53.105+0000 I  COMMAND  [conn231] command baemin.orders command: find { find: "orders", filter: { $or: [ { status: "READY" }, { status: "ASSIGNED" }, { status: "PICKED" }, { status: "CANCELED" } ] }, sort: { createdAt: -1 }, projection: { _id: 1 }, skip: 3999, limit: 1, returnKey: false, showRecordId: false, lsid: { id: UUID("ae9a6090-6c5d-4239-a98f-95d4dd6946c3") }, $db: "baemin" } planSummary: IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: 1 }, IXSCAN { status: 1, createdAt: -1 } keysExamined:4003 docsExamined:1 cursorExhausted:1 numYields:72 nreturned:1 queryHash:156AFF6B planCacheKey:52274FF3 reslen:149 locks:{ ReplicationStateTransition: { acquireCount: { w: 73 } }, Global: { acquireCount: { r: 73 } }, Database: { acquireCount: { r: 73 } }, Collection: { acquireCount: { r: 73 } }, Mutex: { acquireCount: { r: 1 } } } storage:{} protocol:op_query 28ms
```

Reply:
I thought your workaround is cool. 
"Less than totalLimit" case execute 1 more query, but if "Less than totalLimit" mean not much documents, it will be not bad performance.
(And i thought mongodb engine also will be using cached before query result.... I hope...)

Reply:
Your workaround seems good. But I am just thinking about the test result.
If **countDocuments** runs very fast (5ms), should we apply for both cases? Look like it can be faster than **.skip(totalLimit - 1).limit(1)**. If so, we don't need to check both cases and use only **countDocuments** like: {{Math.min(totalLimit, await this.countDocuments(query))}}. Just my opinion because I'm not sure if in case of **Greater than totalLimit**, countDocuments is faster than limit or not.

Reply:
Thank you, guys.

@nhatphongdo
One of the reasons for limiting the total items was the performance issue. When calling the `countDocuments` with large and loose filters, especially for the orders, it needs to traverse too many documents even if it indexed.

It looks like this is the nature of the MongoDB itself, and I guess that the reason why many people requests the COUNT_SCAN to use is the slowness of the counting.

My actual final shape of the goal Is not sending a total count at all for pagination. But it will make a lot of works for every client.

```
db.orders.aggregate([ { $match: { $or: [ { status: "ACCEPTED" }, { status: "PICKED" }, { status: "DELIVERED" } ] } }, { $group: { _id: null, n: { $sum: 1 } } } ])
// mongo_1  | 2020-04-16T07:43:32.509+0000 I  COMMAND  [conn254] command baemin.orders command: aggregate { aggregate: "orders", pipeline: [ { $match: { $or: [ { status: "ACCEPTED" }, { status: "PICKED" }, { status: "DELIVERED" } ] } }, { $group: { _id: null, n: { $sum: 1.0 } } } ], cursor: {}, $db: "baemin", $readPreference: { mode: "primaryPreferred" } } planSummary: IXSCAN { status: 1, statusUpdatedAt: -1 }, IXSCAN { status: 1, createdAt: 1 } keysExamined:1884942 docsExamined:0 cursorExhausted:1 numYields:14731 nreturned:1 queryHash:F75B98C2 planCacheKey:3A63C8A9 reslen:126 locks:{ ReplicationStateTransition: { acquireCount: { w: 14733 } }, Global: { acquireCount: { r: 14733 } }, Database: { acquireCount: { r: 14733 } }, Collection: { acquireCount: { r: 14733 } }, Mutex: { acquireCount: { r: 2 } } } storage:{ data: { bytesRead: 2721, timeReadingMicros: 589 } } protocol:op_msg 2989ms
db.orders.count({ $or: [ { status: "ACCEPTED" }, { status: "PICKED" }, { status: "DELIVERED" } ] })
// mongo_1  | 2020-04-16T07:45:43.768+0000 I  COMMAND  [conn253] command baemin.orders appName: "MongoDB Shell" command: count { count: "orders", query: { $or: [ { status: "ACCEPTED" }, { status: "PICKED" }, { status: "DELIVERED" } ] }, fields: {}, lsid: { id: UUID("263e173e-1c4e-406b-b58f-e94a3d877f41") }, $db: "baemin" } planSummary: IXSCAN { status: 1, statusUpdatedAt: -1 }, IXSCAN { status: 1, createdAt: -1 } keysExamined:1884941 docsExamined:0 numYields:14731 queryHash:F75B98C2 planCacheKey:3A63C8A9 reslen:45 locks:{ ReplicationStateTransition: { acquireCount: { w: 14732 } }, Global: { acquireCount: { r: 14732 } }, Database: { acquireCount: { r: 14732 } }, Collection: { acquireCount: { r: 14732 } }, Mutex: { acquireCount: { r: 1 } } } storage:{} protocol:op_msg 2846ms
db.orders.countDocuments({ $or: [ { status: "ACCEPTED" }, { status: "PICKED" }, { status: "DELIVERED" } ] })
// mongo_1  | 2020-04-16T07:46:43.675+0000 I  COMMAND  [conn253] command baemin.orders appName: "MongoDB Shell" command: aggregate { aggregate: "orders", pipeline: [ { $match: { $or: [ { status: "ACCEPTED" }, { status: "PICKED" }, { status: "DELIVERED" } ] } }, { $group: { _id: null, n: { $sum: 1.0 } } } ], cursor: {}, lsid: { id: UUID("263e173e-1c4e-406b-b58f-e94a3d877f41") }, $db: "baemin" } planSummary: IXSCAN { status: 1, statusUpdatedAt: -1 }, IXSCAN { status: 1, createdAt: -1 } keysExamined:1884942 docsExamined:0 cursorExhausted:1 numYields:14731 nreturned:1 queryHash:F75B98C2 planCacheKey:3A63C8A9 reslen:126 locks:{ ReplicationStateTransition: { acquireCount: { w: 14733 } }, Global: { acquireCount: { r: 14733 } }, Database: { acquireCount: { r: 14733 } }, Collection: { acquireCount: { r: 14733 } }, Mutex: { acquireCount: { r: 2 } } } storage:{} protocol:op_msg 2774ms
db.orders.find({ $or: [ { status: "ACCEPTED" }, { status: "PICKED" }, { status: "DELIVERED" } ] }).count()
// mongo_1  | 2020-04-16T07:47:54.692+0000 I  COMMAND  [conn253] command baemin.orders appName: "MongoDB Shell" command: count { count: "orders", query: { $or: [ { status: "ACCEPTED" }, { status: "PICKED" }, { status: "DELIVERED" } ] }, fields: {}, lsid: { id: UUID("263e173e-1c4e-406b-b58f-e94a3d877f41") }, $db: "baemin" } planSummary: IXSCAN { status: 1, statusUpdatedAt: -1 }, IXSCAN { status: 1, createdAt: 1 } keysExamined:1884942 docsExamined:0 numYields:14731 queryHash:F75B98C2 planCacheKey:3A63C8A9 reslen:45 locks:{ ReplicationStateTransition: { acquireCount: { w: 14732 } }, Global: { acquireCount: { r: 14732 } }, Database: { acquireCount: { r: 14732 } }, Collection: { acquireCount: { r: 14732 } }, Mutex: { acquireCount: { r: 1 } } } storage:{ data: { bytesRead: 18773024, timeReadingMicros: 34581 } } protocol:op_msg 2747ms
db.orders.find({ $or: [ { status: "ACCEPTED" }, { status: "PICKED" }, { status: "DELIVERED" } ] }).hint({ status:1 }).count()
// mongo_1  | 2020-04-16T07:50:03.115+0000 I  COMMAND  [conn253] command baemin.orders appName: "MongoDB Shell" command: count { count: "orders", query: { $or: [ { status: "ACCEPTED" }, { status: "PICKED" }, { status: "DELIVERED" } ] }, hint: { status: 1.0 }, fields: {}, lsid: { id: UUID("263e173e-1c4e-406b-b58f-e94a3d877f41") }, $db: "baemin" } planSummary: IXSCAN { status: 1 } keysExamined:1884942 docsExamined:0 numYields:14726 reslen:45 locks:{ ReplicationStateTransition: { acquireCount: { w: 14727 } }, Global: { acquireCount: { r: 14727 } }, Database: { acquireCount: { r: 14727 } }, Collection: { acquireCount: { r: 14727 } }, Mutex: { acquireCount: { r: 1 } } } storage:{ data: { bytesRead: 7685168, timeReadingMicros: 6134 } } protocol:op_msg 1937ms
```

Reply:
I understood now. Thanks for your very detailed explanation.